tls/s2n_x509_validator.c

```diff
@@ -486,6 +489,69 @@ static S2N_RESULT s2n_x509_validator_process_cert_chain(struct s2n_x509_validato
     return S2N_RESULT_OK;
 }
 
+static S2N_RESULT s2n_x509_validator_set_no_check_time_flag(struct s2n_x509_validator *validator)
+{
+    RESULT_ENSURE_REF(validator);
+    RESULT_ENSURE_REF(validator->store_ctx);
+
+    X509_VERIFY_PARAM *param = X509_STORE_CTX_get0_param(validator->store_ctx);
+    RESULT_ENSURE_REF(param);
+
+#ifdef S2N_LIBCRYPTO_SUPPORTS_FLAG_NO_CHECK_TIME
+    RESULT_GUARD_OSSL(X509_VERIFY_PARAM_set_flags(param, X509_V_FLAG_NO_CHECK_TIME),
+            S2N_ERR_INTERNAL_LIBCRYPTO_ERROR);
+#else
+    RESULT_BAIL(S2N_ERR_UNIMPLEMENTED);
+#endif
+
+    return S2N_RESULT_OK;
+}
+
+int s2n_disable_validity_period_validation_ossl_verify_callback(int default_ossl_ret, X509_STORE_CTX *ctx)
+{
+    int err = X509_STORE_CTX_get_error(ctx);
+    switch (err) {
+        case X509_V_ERR_CERT_NOT_YET_VALID:
+        case X509_V_ERR_CERT_HAS_EXPIRED:
+            return OSSL_VERIFY_CALLBACK_IGNORE_ERROR;
+        default:
+            break;
+    }
+
+    /* If CRL validation is enabled, the CRL verify callback is set on the X509_STORE_CTX. If the
+     * validity period verify callback is set on the CTX, this will override the CRL verify
+     * callback. For simplicity, the CRL verify callback is called in addition to the validity
+     * period verify callback, ensuring both features can be enabled simultaneously.
+     */
+    return s2n_crl_ossl_verify_callback(default_ossl_ret, ctx);
```

Comment:
I had to look up the definition of s2n_crl_ossl_verify_callback to remember that it doesn't actually do crl validation, and instead disables crl validation so that we can do our own validation later. Maybe that callback could use a more descriptive name, but you can probably just rework this comment a little to make it clear why this is a safe simplification:

"Setting the validity period verify callback will override the CRL verify callback.
We can work around this issue by manually triggering the CRL verify callback.
   
We can trigger our CRL verify callback even when not performing CRL validation because the callback just ignores CRL errors the way this callback ignores timestamp errors: our actual CRL validation is performed later. If CRL validation isn't enabled, our CRL verify callback is a no-op."